internal/experimental/registry/client.go

```diff
@@ -17,325 +17,476 @@ limitations under the License.
-		oras.WithConfig(*r.Config), oras.WithNameValidation(nil))
+)
+
+// Login logs into a registry
+func (c *Client) Login(host string, options ...LoginOption) (*LoginResult, error) {
+	operation := &loginOperation{}
+	for _, option := range options {
+		option(operation)
+	}
+	authorizerLoginOpts := []auth.LoginOption{
+		auth.WithLoginContext(ctx(c.out, c.debug)),
+		auth.WithLoginHostname(host),
+		auth.WithLoginUsername(operation.username),
+		auth.WithLoginSecret(operation.password),
+		auth.WithLoginUserAgent(version.GetUserAgent()),
+	}
+	if operation.insecure {
+		authorizerLoginOpts = append(authorizerLoginOpts, auth.WithLoginInsecure())
+	}
+	err := c.authorizer.LoginWithOpts(authorizerLoginOpts...)
 	if err != nil {
-		return err
+		return nil, err
+	}
+	result := &LoginResult{
+		Host: host,
 	}
-	s := ""
-	numLayers := len(layers)
-	if 1 < numLayers {
-		s = "s"
+	fmt.Fprintln(c.out, "Login Succeeded")
+	return result, nil
+}
+
+// LoginOptBasicAuth returns a function that sets the username/password settings on login
+func LoginOptBasicAuth(username string, password string) LoginOption {
+	return func(operation *loginOperation) {
+		operation.username = username
+		operation.password = password
 	}
-	fmt.Fprintf(c.out,
-		"%s: pushed to remote (%d layer%s, %s total)\n", r.Tag, numLayers, s, byteCountBinary(r.Size))
-	return nil
 }
 
-// PullChart downloads a chart from a registry
-func (c *Client) PullChart(ref *Reference) (*bytes.Buffer, error) {
-	buf := bytes.NewBuffer(nil)
+// LoginOptInsecure returns a function that sets the insecure setting on login
+func LoginOptInsecure(insecure bool) LoginOption {
+	return func(operation *loginOperation) {
+		operation.insecure = insecure
+	}
+}
 
-	if ref.Tag == "" {
-		return buf, errors.New("tag explicitly required")
+type (
+	// LogoutOption allows specifying various settings on logout
+	LogoutOption func(*logoutOperation)
+
+	// LogoutResult is the result returned upon successful logout.
+	LogoutResult struct {
+		Host string `json:"host"`
 	}
 
-	fmt.Fprintf(c.out, "%s: Pulling from %s\n", ref.Tag, ref.Repo)
+	logoutOperation struct{}
+)
 
-	store := content.NewMemoryStore()
-	fullname := ref.FullName()
-	_ = fullname
-	_, layerDescriptors, err := oras.Pull(ctx(c.out, c.debug), c.resolver, ref.FullName(), store,
-		oras.WithPullEmptyNameAllowed(),
-		oras.WithAllowedMediaTypes(KnownMediaTypes()))
+// Logout logs out of a registry
+func (c *Client) Logout(host string, opts ...LogoutOption) (*LogoutResult, error) {
+	operation := &logoutOperation{}
+	for _, opt := range opts {
+		opt(operation)
+	}
+	err := c.authorizer.Logout(ctx(c.out, c.debug), host)
 	if err != nil {
-		return buf, err
+		return nil, err
 	}
-
-	numLayers := len(layerDescriptors)
-	if numLayers < 1 {
-		return buf, errors.New(
-			fmt.Sprintf("manifest does not contain at least 1 layer (total: %d)", numLayers))
+	result := &LogoutResult{
+		Host: host,
```

Comment:
It looks like the host is just passed through the function.  The host is already known if it is a param.

Reply:
Yes this is true. I simply wanted to populate the result with some information

Reply:
Right. but why would the caller need a return type with the same host name they passed as a function parameter?

Reply:
Resolved in 8202ccd6998e6bf8aa34c8d9d7527e08d326397b